cmd/dp-code-list-api/main.go

```diff
@@ -5,92 +5,155 @@ import (
 	httpServer := server.New(cfg.BindAddr, router)
 	httpServer.HandleOSSignals = false
 
-	healthTicker := healthcheck.NewTicker(
-		cfg.HealthCheckInterval,
-		cfg.HealthCheckRecovery,
-		datastore,
-	)
+	// Graceful shutdown function
+	shutdown := func() {
 
-	shutdown := func(httpShutdown bool) {
-		log.Info(fmt.Sprintf("shutdown with timeout: %d", cfg.GracefulShutdownTimeout), nil)
-		ctx, cancel := context.WithTimeout(context.Background(), cfg.GracefulShutdownTimeout)
+		shutdownCtx, cancel := context.WithTimeout(context.Background(), cfg.GracefulShutdownTimeout)
+		anyError := false
+		log.Event(shutdownCtx, "shutdown with timeout", log.INFO, log.Data{"timeout": cfg.GracefulShutdownTimeout})
 
-		if httpShutdown {
-			err = httpServer.Shutdown(ctx)
-			if err != nil {
-				log.Error(err, nil)
-			}
+		// StopHTTP Server
+		err = httpServer.Shutdown(shutdownCtx)
+		if err != nil {
+			anyError = true
+			log.Event(shutdownCtx, "http server shutdown error", log.ERROR, log.Error(err))
+		} else {
+			log.Event(shutdownCtx, "http server successful shutdown", log.INFO)
 		}
 
-		healthTicker.Close()
+		// Strop healthcheck
+		hc.Stop()
+		log.Event(shutdownCtx, "healthcheck stopped", log.INFO)
 
-		if err = datastore.Close(ctx); err != nil {
-			log.Error(err, nil)
+		// Close data store
+		if err = datastore.Close(shutdownCtx); err != nil {
+			anyError = true
+			log.Event(shutdownCtx, "datastore close error", log.ERROR, log.Error(err))
+		} else {
+			log.Event(shutdownCtx, "datastore successfully closed", log.INFO)
 		}
 
-		log.Info("shutdown complete", nil)
-
+		// cancel the timer in the shutdown context.
 		cancel()
-		os.Exit(1)
+
+		// if any error happened during shutdown, log it and exit with err code
+		if anyError {
+			log.Event(ctx, "graceful shutdown had errors", log.WARN)
+			os.Exit(1)
+		}
+
+		// if all dependencies shutted down successfully, log it and exit with success code
+		log.Event(ctx, "graceful shutdown was successful", log.INFO)
+		os.Exit(0)
 	}
 
+	// Start healthcheck ticker
+	hc.Start(ctx)
+
+	// Start HTTP Server
 	go func() {
-		log.Info("code list api starting.....", log.Data{"bind_addr": cfg.BindAddr})
+		log.Event(ctx, "code list api starting.....", log.INFO, log.Data{"bind_addr": cfg.BindAddr})
 		if err := httpServer.ListenAndServe(); err != nil && err != http.ErrServerClosed {
-			log.Error(err, nil)
+			hc.Stop()
+			log.Event(ctx, "error starting http server", log.ERROR, log.Error(err))
 			httpErrChannel <- err
 		} else {
 			httpErrChannel <- errors.New("http completed shutdown")
 		}
 	}()
 
-	for {
-		select {
-		case err := <-httpErrChannel:
-			log.ErrorC("http error received", err, nil)
-			shutdown(false)
-		case <-signals:
-			log.Debug("os signal received", nil)
-			shutdown(true)
+	// log errors
+	go func() {
+		for {
+			select {
+			case err := <-httpErrChannel:
+				log.Event(ctx, "http error received", log.ERROR, log.Error(err))
```

Comment:
I don't think we need this separate go routine now. All it appears to do is log an error message from the HTTP server. The go routine above already does this? I think we can remove this go routine and the `httpErrChannel`

Reply:
Yes, I think you're right. I think this channel existed in the past because we wanted to shutdown on HTTP server error.

As we no longer require to do this, and we log the errors in the HTTP server goroutine anyway, you are right.